csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,625 @@ bool predicateAtEnd(ForLoop* loop) {
+    // then these two groups together represents linear_g.
+    NVF_ERROR(to.size() == 2);
+    NVF_ERROR(from.front() == current);
+    return Composition<LinearGroupProjection>{to.front(), to.back()};
+  } else {
+    // If linear_g is merged with another group, then part of the merged group
+    // represents linear_g.
+    NVF_ERROR(from.size() == 2);
+    NVF_ERROR(to.size() == 1);
+    NVF_ERROR(from.front() == current || from.back() == current);
+    return PartOf<LinearGroupProjection>{
+        std::make_shared<LinearGroupProjection>(to.front()),
+        /*inner_extent=*/from.front() == current
+            ? from.back()->front()->as<IterDomain>()->extent()
+            : nullptr,
+        /*selected_extent=*/
+        simplifyExpr(current->front()->as<IterDomain>()->extent())};
+  }
+}
+
+LinearGroupProjection propagate(
+    const PartOf<LinearGroupProjection>& current,
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  auto from = fromGroups(id_graph, eg, direction);
+  auto to = toGroups(id_graph, eg, direction);
+  if (from.size() == 1) {
+    NVF_ERROR(to.size() == 2);
+    NVF_ERROR(related(*current.group, from.front()));
+    auto group = propagate(*current.group, id_graph, eg, direction);
+    if (!group.hasValue()) {
+      // Propagation of group may fail. For example, if the group before
+      // propagation is [4, 3], and eg is split(3, 2), then the end result
+      // would be like [4, 2, 2]. However, because the split is indivisible,
+      // this makes group discontiguous. That is, "a composition of 4 and 3"
+      // is not a "composition of 4, 2, and 2". In this case, the return value
+      // of `propagate` will be std::monostate, indicating that the linear
+      // proving fails. For this case, we just propagate the failure.
+      return {};
+    }
+    // The result is actually just the
+    //   PartOf{group, current.inner_extent, current.selected_extent}
+    // However, we want to simplify the result. For example, if the inner_extent
+    // is a multiple of the extent of the last item in group, we can simplify
+    // the inner_extent and group by canceling the last item in group.
+    Val* new_inner_extent = current.inner_extent;
+    Val* group_extent = nullptr;
+    // Now we will simplify inner_extent and `group`, for example, by removing
+    // common factors. The simplified group will be updated in the variable
+    // `group`, the extent of the group after simplification is stored in
+    // `group_extent`, and the new inner_extent after simplification is stored
+    // in `new_inner_extent`.
+    if (group.is<Composition>()) {
+      auto dq = group.as<Composition>();
+      // If the new_inner_extent is divisible by the extent of the last item
+      // in dq, we can simplify new_inner_extent and dq by canceling dq's last
+      // items. For example, if new_inner_extent is 6 and dq is [5, 3], we can
+      // simplify them to 2 and [5].
```

Comment:
I'm puzzled here as I'm not sure why this is always valid simplification. 

IIUC, `dq: [5, 3], new_inner_extent: 6` means that there's a 2D tile of shape `5x3` (outer x inner). Since `new_inner_extent` is 6, the projection to this 2D tile is the `[2:, :]` subregion. I'm not sure why the inner axis can be dropped. Suppose the next expression where this projection is propagated along is `split the inner domain by 2`. If `[5, 3]` were simplified to `[5]`, I'm not sure how the next expression should be interpreted as there's no longer the inner domain of extent 3.

Reply:
> the projection to this 2D tile is the `[2:, :]` subregion

Do you mean `[::2, 0]`?
```
>>> t = torch.arange(15).view(5, 3)
tensor([[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8],
        [ 9, 10, 11],
        [12, 13, 14]])
>>> t[::2, :]
tensor([[ 0,  1,  2],
        [ 6,  7,  8],
        [12, 13, 14]])
```

> Suppose the next expression where this projection is propagated along is `split the inner domain by 2`. If `[5, 3]` were simplified to `[5]`, I'm not sure how the next expression should be interpreted as there's no longer the inner domain of extent 3.

`[5, 3]` is simplified as `[5]`, and `new_inner_extent` will become 2. When we see `split the inner domain by 2`, we will get
```C++
PartOf{group=[3, 2], inner_extent=2}
```
which will further be simplified as
```C++
PartOf{group=[3], inner_extent=1}
```
which will further be simplified as `[3]`, so `linear_g` is projected as the outer of the split.

This result makes sense, because, for example, if the `[5, 3]` is:
```
 0  1  2
 3  4  5
 6  7  8
 9 10 11
12 13 14
```

Then `PartOf{group=[5, 3], inner_extent=6}` means pad and view [5, 3] as [3, 6]:
```
 0  1  2  3  4  5
 6  7  8  9 10 11
12 13 14  _  _  _
```

Canceling the 3 just means: let's always consider each 3 items as a whole. Let's use `()` to denote a whole item that can not be further divided, then cancelling the 3 in `[5, 3]` as `[5]` means considering the above `[5, 3]` as

```
x1 = ( 0  1  2)
x2 = ( 3  4  5)
x3 = ( 6  7  8)
x4 = ( 9 10 11)
x5 = (12 13 14)
```

and `PartOf{group=[5], inner_extent=2}` means pad and view the [5] as [3, 2]:

```
x1 x2
x3 x4
x5  _
```

Splitting the [5] as [3, 2] means pad and view the same way as above, and `PartOf{group=[3, 2], inner_extent=2}` means viewing the above [3, 2] tensor as is because the inner is already 2.

Cancelling the 2 in `[3, 2]` as [3] means considering every two items as a whole, that is:

```
y1 = (x1 x2)
y2 = (x3 x4)
y3 = (x5  _)
```